<a href="https://colab.research.google.com/github/Shakilkhan24/Playground_DL/blob/main/playing_HF_custom_data_torch_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!pip install transformers
!pip install datasets



```
# Conversion of HF data to pytorch
```





In [ ]:
from datasets import load_dataset
data=load_dataset('glue','mrpc')

# Working with HuggingFace DATAS

In [25]:
from datasets import load_dataset
dataset2 = load_dataset("glue", "mrpc", split="train")

In [37]:
print(data)
print('___________seperator____________')
print(dataset2)    # only picking train datas

DatasetDict({
    train: Dataset({
        features: ['sentence1', 'sentence2', 'label', 'idx'],
        num_rows: 3668
    })
    validation: Dataset({
        features: ['sentence1', 'sentence2', 'label', 'idx'],
        num_rows: 408
    })
    test: Dataset({
        features: ['sentence1', 'sentence2', 'label', 'idx'],
        num_rows: 1725
    })
})
___________seperator____________
Dataset({
    features: ['sentence1', 'sentence2', 'label', 'idx'],
    num_rows: 3668
})




```
# casting dataset
data.cast()
```



In [41]:
print('dataset_features before',dataset2.features)
print('_____________________')
print('_____________________')

from datasets import ClassLabel, Value
new_features = dataset2.features.copy()
new_features["label"] = ClassLabel(names=["negative", "positive"])
new_features["idx"] = Value("int64")
dataset2 = dataset2.cast(new_features)
print('dataset after modifying the features types')
print('______________________________')
dataset2.features

dataset_features before {'sentence1': Value(dtype='string', id=None), 'sentence2': Value(dtype='string', id=None), 'label': ClassLabel(names=['not_equivalent', 'equivalent'], id=None), 'idx': Value(dtype='int32', id=None)}
_____________________
_____________________


Casting the dataset:   0%|          | 0/3668 [00:00<?, ? examples/s]

dataset after modifying the features types
______________________________


{'sentence1': Value(dtype='string', id=None),
 'sentence2': Value(dtype='string', id=None),
 'label': ClassLabel(names=['negative', 'positive'], id=None),
 'idx': Value(dtype='int64', id=None)}

In [27]:
data=data.shuffle(seed=42)

# we can shuffle the datas before procedding to training loop


Image DataSet from Hugging_Face ...


```
Look like interactible with the data format .
We can easily preprocess using pytorch or others.
```





In [52]:
from datasets import Image

dataset = load_dataset("beans", split="train")    # image dataset here

In [5]:
print(dataset)     # work on this

Dataset({
    features: ['image_file_path', 'image', 'labels'],
    num_rows: 1034
})


In [32]:
print(len(dataset[0]))
print(dataset[0].keys())

3
dict_keys(['image_file_path', 'image', 'labels'])


JUST look at the DATA

In [33]:
import pandas as pd
x=pd.DataFrame(dataset)

In [34]:
x.head()

image_file_path  \
0  /home/albert/.cache/huggingface/datasets/downl...   
1  /home/albert/.cache/huggingface/datasets/downl...   
2  /home/albert/.cache/huggingface/datasets/downl...   
3  /home/albert/.cache/huggingface/datasets/downl...   
4  /home/albert/.cache/huggingface/datasets/downl...   

                                               image  labels  
0  <PIL.JpegImagePlugin.JpegImageFile image mode=...       0  
1  <PIL.JpegImagePlugin.JpegImageFile image mode=...       0  
2  <PIL.JpegImagePlugin.JpegImageFile image mode=...       0  
3  <PIL.JpegImagePlugin.JpegImageFile image mode=...       0  
4  <PIL.JpegImagePlugin.JpegImageFile image mode=...       0



```
# BUT REMEMBER , IT'S THE PYTHON FUNCTIONS, NOT FROM 'PANDAS' LIBARY
dataset.map(any_func)
dataset.filter(lambda x: something(x))
dataset.apply()
or more functions you can use
```



In [36]:
# Many functions are associated with the Dataset object from hugging face ...
# for example dataset.rename_columns,remove_columns,select_columns and other stuffs ,

# KIND OF PANDAS DATAFRAME LIKE



In [ ]:
# modified_data=dataset.filter(lambda x: x['image'])

# x contains three values , in this case lambda function only returing image
#  meaning here you can look your datas in different ways




```
# dataset.train_test_split(test_size=0.2)
Flavor of scikit-learn train_test_split module
```



You can look your datas in different ways...

In [28]:
sub_set_data=dataset.filter(lambda example:example['labels']==0)

Filter:   0%|          | 0/1034 [00:00<?, ? examples/s]

In [45]:
# Ways of preprocessing data based on any specific requirements

def augment_data(examples):
    outputs = []
    for sentence in examples["sentence1"]:
        words = sentence.split(' ')
        # preprocess the data , whatever requirements you have
        pass





In [46]:
from transformers import AutoTokenizer
tokenizer=AutoTokenizer.from_pretrained('bert-base-uncased')

tokenizer_config.json:   0%|          | 0.00/28.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]



```
# PREPROCESSING IS NOT HARD PROBABLY
```



In [47]:
y=dataset2.map(lambda example:tokenizer([example['sentence1'],example['sentence2']],truncation=True,padding=True,return_tensors='pt'))

Map:   0%|          | 0/3668 [00:00<?, ? examples/s]

In [49]:
print(y)
print(y[0].keys())

Dataset({
    features: ['sentence1', 'sentence2', 'label', 'idx', 'input_ids', 'token_type_ids', 'attention_mask'],
    num_rows: 3668
})
dict_keys(['sentence1', 'sentence2', 'label', 'idx', 'input_ids', 'token_type_ids', 'attention_mask'])


In [29]:
print('datas with label=0',len(sub_set_data))
print('datas with label=1',len(dataset)-len(sub_set_data))


datas with label=0 345
datas with label=1 689


In [6]:
print(dataset[0])

{'image_file_path': '/home/albert/.cache/huggingface/datasets/downloads/extracted/967f0d9f61a7a8de58892c6fab6f02317c06faf3e19fba6a07b0885a9a7142c7/train/angular_leaf_spot/angular_leaf_spot_train.0.jpg', 'image': <PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=500x500 at 0x7DC2006EA920>, 'labels': 0}


In [19]:
first_image = dataset[0]['image']
first_image.show()

In [17]:
# import matplotlib.pyplot as plt
# from PIL import Image
# path=dataset[0]['image_file_path']
# x=Image.open(path)
# plt.imshow(x)
# plt.show()

In [21]:
path_of_image=dataset[0]['image_file_path']
print(path_of_image)

/home/albert/.cache/huggingface/datasets/downloads/extracted/967f0d9f61a7a8de58892c6fab6f02317c06faf3e19fba6a07b0885a9a7142c7/train/angular_leaf_spot/angular_leaf_spot_train.0.jpg


WAY OF PICKING DATA_FOLDER FROM HUGGING FACE DATA HUB...

In [ ]:
folder_path='/content/drive/MyDrive/data_set/classifier_data/Animals'

custom_data = load_dataset("Folder", data_dir=folder_path)

# There is an error, becuase we can only load data from hugging face hub
# Not our custom data using the load_dataset_module


# pytorch custom data class



```
# example contains
dict_keys(['sentence1', 'sentence2', 'label', 'idx', 'input_ids',
'token_type_ids', 'attention_mask'])
```



TEXT DATA

In [50]:
import torch
from datasets import load_dataset
from transformers import DistilBertTokenizer, DistilBertForSequenceClassification
from torch.utils.data import DataLoader, Dataset
from transformers import Trainer, TrainingArguments

# Load dataset
dataset = load_dataset("glue", "mrpc")

# Load pre-trained tokenizer and model
tokenizer = DistilBertTokenizer.from_pretrained("distilbert-base-uncased")
model = DistilBertForSequenceClassification.from_pretrained("distilbert-base-uncased", num_labels=2)

# Tokenize and preprocess dataset
def tokenize_function(example):
    return tokenizer(example["sentence1"], example["sentence2"], truncation=True)

tokenized_dataset = dataset.map(tokenize_function, batched=True)

# Convert tokenized dataset to PyTorch Dataset
class TorchDataset(Dataset):
    def __init__(self, examples):
        self.examples = examples

    def __len__(self):
        return len(self.examples["input_ids"])

    def __getitem__(self, idx):
        return {
            "input_ids": self.examples["input_ids"][idx],
            "attention_mask": self.examples["attention_mask"][idx],
            "labels": self.examples["label"][idx]
        }

train_dataset = TorchDataset(tokenized_dataset["train"])
eval_dataset = TorchDataset(tokenized_dataset["validation"])

train_dataloader = DataLoader(train_dataset, batch_size=8, shuffle=True)
eval_dataloader = DataLoader(eval_dataset, batch_size=8)

# # Define training arguments
# training_args = TrainingArguments(
#     per_device_train_batch_size=8,
#     per_device_eval_batch_size=8,
#     num_train_epochs=3,
#     logging_dir='./logs',
#     logging_steps=100,
# )

# # Define Trainer
# trainer = Trainer(
#     model=model,
#     args=training_args,
#     train_dataset=train_dataset,
#     eval_dataset=eval_dataset,
# )

# # Train the model
# trainer.train()

# # Evaluate the model
# trainer.evaluate()


tokenizer_config.json:   0%|          | 0.00/28.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/483 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

Some weights of DistilBertForSequenceClassification were not initialized from the model checkpoint at distilbert-base-uncased and are newly initialized: ['classifier.weight', 'pre_classifier.bias', 'classifier.bias', 'pre_classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Map:   0%|          | 0/3668 [00:00<?, ? examples/s]

Map:   0%|          | 0/408 [00:00<?, ? examples/s]

Map:   0%|          | 0/1725 [00:00<?, ? examples/s]

IMAGE DATA

Dataset({
    features: ['image_file_path', 'image', 'labels'],
    num_rows: 1034
})

In [ ]:
import torch
from datasets import load_dataset
from torchvision import transforms
from PIL import Image
from torch.utils.data import Dataset

class CustomImageDataset(Dataset):
    def __init__(self, dataset_name, split_name, transform=None):
        self.dataset = load_dataset(dataset_name, split=split_name)
        self.transform = transform

    def __len__(self):
        return len(self.dataset)

    def __getitem__(self, idx):
        image_path = self.dataset[idx]['image_file_path']
        image = Image.open(image_path)
        label = self.dataset[idx]['label']

        if self.transform:
            image = self.transform(image)

        return image, label

# Define transforms for preprocessing
transform = transforms.Compose([
    transforms.Resize((224, 224)),  # Resize the image to 224x224
    transforms.ToTensor(),           # Convert PIL Image to tensor
    transforms.Normalize(            # Normalize image tensor
      mean=[0.485, 0.456, 0.406],   # Mean values for RGB channels
      std=[0.229, 0.224, 0.225]     # Standard deviations for RGB channels
    )
])

# Create custom dataset instance
dataset = CustomImageDataset(dataset_name="beans", split_name="train", transform=transform)

# Example usage
# Accessing a sample image and label
sample_image, sample_label = dataset[0]
print("Sample image shape:", sample_image.shape)
print("Sample label:", sample_label)


In [53]:
print(dataset)

Dataset({
    features: ['image_file_path', 'image', 'labels'],
    num_rows: 1034
})
